fix: properly expand absolute path placeholder for link calls

voxeljorge opened the pull request:
**What type of PR is this?**

> Bug fix

**What does this PR do? Why is it needed?**

PR #4009 added a mechanism that uses a sentinel value as a placeholder for paths which need to be expanded to absolutes for compile operations. These paths end up embedded in `.a` files and come back to the linker during link operations. The previous PR replaces the `CC` tool with a pointer to `builder cc` which in turn substitutes the placeholder value for actual paths. Unfortunately this wasn't done for the linker, so the linker is receiving parameters like `--sysroot=__GO_BAZEL_CC_PLACEHOLDER__external/_main~_repo_rules~ubuntu_20_04_sysroot/` which is not valid. When a sysroot is configured this means that internally the linker won't be able to find the sysroot in some cases.

In almost comically bad luck, this *only* affects the `go tool link` calls to `linkerFlagSupported` because for some reason `linkerFlagSupported` puts the `-extldflags` at the *beginning* of the linker command, while the main host link command has them at the end. So this means the linker detects all flags as unsupported but then proceeds to run the linker "correctly" which results in really confusing linker errors. In my case this resulted in the linker trying to run with `-rdynamic` incorrectly.

**Which issues(s) does this PR fix?**

Fixes #3886 

It could be reasonable to open a bug against `go tool link` describing the difference between the host link and the flag checks. I suspect it's not really intended behavior and was just sort of happenstance that things ended up that way.

I also really have no idea how to write a test for this. I'm willing to give it a shot but would really appreciate a suggestion as to where/how.

voxeljorge commented:
Clearly this naive approach won't work on windows, I guess we could detect the OS write an OS specific script? This all feels pretty messy so i'm hoping someone out there has ideas for a better solution.

voxeljorge commented on go/tools/builders/cc.go:
```diff
@@ -13,12 +13,13 @@ import (
 func cc(args []string) error {
 	cc := os.Getenv("GO_CC")
 	if cc == "" {
-		errors.New("GO_CC environment variable not set")
```
this looks like just an error in the original code, `errors.New`here did nothing and was effectively just ignoring the error.

fmeum replied:
Nice catch. I think that you can just return the error here instead. I wonder why nogo didn't catch this.

voxeljorge replied:
fixed

voxeljorge commented:
One idea for a different approach would be to use a symlink named something like `builder-cc` and have the builder inspect `os.Argv[0]` to set the command.

voxeljorge commented:
I've changed this to use a symlink approach instead of a script which seems to be more portable.

fmeum commented on go/tools/builders/link.go:
```diff
@@ -141,6 +141,16 @@ func link(args []string) error {
 
+	// substitute `builder cc` for the linker with a script. unfortunately
+	// we can't just set an environment variable to `builder cc` because
+	// in `go tool link` the `linkerFlagSupported` call sites used to determine
```
Could you add some references to the source code here? That would make it easier to look for alternative solutions.

voxeljorge replied:
I'm not quite sure if you want references to the code in the comment above or just here in github but here's the relevant line:

https://cs.opensource.google/go/go/+/master:src/cmd/link/internal/ld/lib.go;l=1732?q=hostlink

You can see the call to `linkerFlagSupported()` uses `argv[0]` as the `linker`. `argv` is set up here:

https://cs.opensource.google/go/go/+/master:src/cmd/link/internal/ld/lib.go;drc=f2d118fd5f7e872804a5825ce29797f81a28b0fa;l=479?q=hostlink

This uses the global `flagExtLd` which is a `quoted.Flag`, a type defined here:

https://cs.opensource.google/go/go/+/master:src/cmd/internal/quoted/quoted.go;drc=f2d118fd5f7e872804a5825ce29797f81a28b0fa;l=107

This type is a `[]string` which means that `flag.Var` will parse this value into a slice of strings, resulting in something like `[]string{"/path/to/builder", "cc"}`.

fmeum replied:
Could you add the first two to the comment? Thanks for looking them up!

voxeljorge replied:
np, done

fmeum reviewed (commented):
Thanks for picking this up, it looks genuinely tricky. I'll think about the approach.

Could you file an upstream issue and link it here? I agree that this is pretty clearly unintended behavior and having a fix in Go would allow us to drop complicated workarounds at some point.

fmeum commented:
Could you add a test for this under `tests/core/cgo`? It doesn't have to be a full end-to-end test, but perhaps you could set some `--copts` that are turned into absolute paths and assert on the build output with `-s`? That way you don't have to set up an actual sysroot in the test.

voxeljorge commented:
Sure, I will do both.

I honestly think even if an upstream issue is fixed there is still technically a bug here, we just won't hit the bug for sysroots because the last one will take precedence but it is conceivable that other parameters with paths wouldn't work the same way and we would end up with some new subtle bug.

voxeljorge commented:
It looks like someone else already filed an issue relating to this:

https://github.com/golang/go/issues/69350

voxeljorge commented:
@fmeum I've added a test. I think the test most likely only works on platforms with gcc-ish compilers (I have no idea if this bug really even exists on windows and have no way to test unfortunately) so I added a build constraint for !windows.

voxeljorge commented:
actually it seems the test passes without the build constraint and fails with it so I'll leave it alone for now. I'm assuming the windows test must still use gcc

fmeum commented on go/tools/builders/env.go:
```diff
@@ -162,6 +163,86 @@ func (e *env) runCommandToFile(out, err io.Writer, args []string) error {
+
+	if err := os.Symlink(absPath, builderPath); err != nil {
+		_ = os.RemoveAll(dirname)
+		return func() {}, "", err
```
This cleanup func should remove the directory.

voxeljorge replied:
what I was going for here was that the cleanup function is only needed if the function succeeds. the code is a little inconsistent in that it returns nil above for the cleanup function and non-nil here, but also returns a non-nil err.

what probably makes the most sense is just to make it consistent and return a nil cleanup function if err is non-nil. the directory should always be removed in case of errors otherwise the caller would have to *also* call the cleanup function on an error which doesn't feel right.

i'll make that change, let me know what you think

fmeum reviewed (approved):
Looks good, just two comments.

voxeljorge commented:
I think the build failure is in an unrelated test, maybe a fluke? Either way I'm unable to trigger a retry without pushing an update.

voxeljorge commented:
@fmeum I believe this may be ready. I addressed one comment slightly differently than you asked so I let it unresolved but I think the issue is fixed. Let me know if you need me to do anything else on this one.

fmeum commented:
Thanks for the very thorough execution!
